src/utils/task.rs

```diff
@@ -143,13 +158,96 @@ impl Services {
         TaskMessageBox(tx)
     }
 
+    /// Spawn the given Future in a new dedicated runtime
+    ///
+    /// * utilising one thread only;
+    /// * 2MiB stack size max
+    pub fn spawn_future<F, T>(&mut self, name: &'static str, f: F)
+    where
+        F: FnOnce(TokioServiceInfo) -> T,
+        T: Future<Item = (), Error = ()> + Send + 'static,
+    {
+        let mut runtime = runtime::Builder::new()
+            .keep_alive(None)
+            .core_threads(1)
+            .blocking_threads(1)
+            .stack_size(2 * 1024 * 1024)
+            .name_prefix(name)
+            .build()
+            .unwrap();
+
+        let executor = runtime.executor();
+
+        let now = Instant::now();
+        let future_service_info = TokioServiceInfo {
+            name: name,
+            up_time: now,
+            logger: get_global_logger().new(o!("task" => name.to_owned())),
+            executor: executor,
+        };
+
+        runtime.spawn(f(future_service_info));
+
+        let task = Service::new_runtime(name, runtime, now);
+        self.services.push(task);
+    }
+
+    /// Spawn a tokio service that will await messages and will be executed
+    /// sequentially for every received inputs
+    ///
+    /// * utilising one thread only;
+    /// * 2MiB stack size max
+    ///
+    pub fn spawn_future_with_inputs<F, Msg, T>(
+        &mut self,
+        name: &'static str,
+        mut f: F,
+    ) -> MessageBox<Msg>
+    where
+        F: FnMut(&TokioServiceInfo, Input<Msg>) -> T,
+        F: Send + 'static,
+        Msg: Send + 'static,
+        T: IntoFuture<Item = (), Error = ()> + Send + 'static,
+        <T as futures::IntoFuture>::Future: Send,
+    {
+        let mut runtime = runtime::Builder::new()
+            .keep_alive(None)
+            .core_threads(1)
+            .blocking_threads(1)
+            .stack_size(2 * 1024 * 1024)
+            .name_prefix(name)
+            .build()
+            .unwrap();
+
+        let executor = runtime.executor();
+
+        let now = Instant::now();
+        let future_service_info = TokioServiceInfo {
+            name: name,
+            up_time: now,
+            logger: get_global_logger().new(o!("task" => name.to_owned())),
+            executor: executor,
+        };
+
+        let (msg_box, msg_queue) = async_msg::channel(1000);
+        let future = msg_queue.for_each(move |input| f(&future_service_info, Input::Input(input)));
+
+        runtime.spawn(future);
+
+        let task = Service::new_runtime(name, runtime, now);
+        self.services.push(task);
```

Comment:
Lots of code duplication here; could this method be done in terms of `spawn_future`?

Reply:
that could be a solution, though there are requirements we still need to comply to regarding some of our services. The leadership task needs to run in a thread and we need to make sure it is well enough isolated from anything else.